### PR TITLE
test: enable branch coverage tracking

### LINKED ISSUE #343: Enable branch coverage in SimpleCov
Currently we're only measuring line coverage, which means one line conditionals like guards are always considered covered even though these often have important behaviour that we should make sure we cover. This also doesn't combine well with our rubocop rules which require us to use one-line guards when possible, which results in "hiding" code from coverage.

https://github.com/simplecov-ruby/simplecov#branch-coverage-ruby--25

### PR DESCRIPTION
**Pull request by G-Rath**
Resolves #343

### CODE REVIEW

**Comment by G-Rath** [comment]
@robotdana @eoinkelly I've updated this to set our minimum coverage for both lines and branches to be 90 as I don't think there's a good reason to favor one or the other, and I think the best way to determine what a good value for `branches` is will be to live with it on a project and get feedback from teams rather than prematurely lowering it since teams won't be spending time monitoring the coverage rate to see if they can _increase_ that value.

Of course out of the box we've only got ~50% branch coverage - I might ask for some help in getting that increased after I've had a chance to look into where our uncovered branches actually are.

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -8,14 +8,15 @@
...
 # Details of default values for these configuration options can be seen at
 # https://github.com/colszowka/simplecov/blob/master/lib/simplecov/configuration.rb#L217
-#
-SimpleCov.minimum_coverage 90
-# SimpleCov.minimum_coverage_by_file 80
-# SimpleCov.maximum_coverage_drop 5
-# SimpleCov.refuse_coverage_drop
-SimpleCov.use_merging false
-
 SimpleCov.start do
```
Comment (robotdana) [inline spec/spec_helper.rb:11]: is there a reason we don't use the built in 'rails' config? this would give us a lot of the filtering up-front

Comment (G-Rath) [reply spec/spec_helper.rb:11]: No idea 🤷

Comment (eoinkelly) [reply spec/spec_helper.rb:11]: TIL about Simplecov profiles (not sure how I've managed to miss them until now :shrug:). The rails profile is https://github.com/simplecov-ruby/simplecov/blob/main/lib/simplecov/profiles/rails.rb and seems useful.

```suggestion
SimpleCov.start("rails") do
```

**Review by eoinkelly — APPROVED** [review]
:sparkles:

**Comment by G-Rath** [comment]
I'm a bit suspect that adding the `rails` profile has caused the test suite to pass - I think we should check that we're not now ignoring some specs.

**Review by robotdana — APPROVED** [review]

**Comment by eoinkelly** [comment]
I built the devise variant using this branch and had a poke around in the `SimpleCov` module with `pry`.

```ruby
  11   │ SimpleCov.start("rails") do
  12   │   enable_coverage :branch
  13   │
  14   │   use_merging false
  15   │   # minimum_coverage_by_file 80
  16   │   # maximum_coverage_drop 5
  17   │   # refuse_coverage_drop
  18   │   minimum_coverage line: 90, branch: 80
  19   │
  20   │   add_filter "/bin/"
  21   │   add_filter "/lib/tasks/auto_annotate_models.rake"
  22   │   add_filter "/lib/tasks/coverage.rake"
  23   │   add_filter "/spec/support/"
  24   │   add_filter "/spec/factories/"
  25   │   add_filter "/spec/rails_helper.rb"
  26   │   add_filter "/spec/spec_helper.rb"
  27   │
  28   │   # ignore files with fewer than 5 lines
  29   │   add_filter do |source_file|
  30   │     source_file.lines.count < 5
  31   │   end
  32   │ end
  33   │
  34 + │ require 'pry'; binding.pry
```


```
[3] pry(main)> cd SimpleCov
[6] pry(SimpleCov):1> @filters
=> [#<SimpleCov::StringFilter:0x0000000109015248 @filter_argument="/vendor/bundle/">,
 #<SimpleCov::RegexFilter:0x0000000109014cd0 @filter_argument=/^\/\..*/>,
 #<SimpleCov::BlockFilter:0x0000000109014758
  @filter_argument=#<Proc:0x0000000109014848 /Users/eoinkelly/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/simplecov-0.21.2/lib/simplecov/profiles/root_filter.rb:6>>,
 #<SimpleCov::StringFilter:0x0000000109043af8 @filter_argument="/test/">,
 #<SimpleCov::StringFilter:0x0000000109043a80 @filter_argument="/features/">,
 #<SimpleCov::StringFilter:0x0000000109043a08 @filter_argument="/spec/">,
 #<SimpleCov::StringFilter:0x0000000109043990 @filter_argument="/autotest/">,
 #<SimpleCov::RegexFilter:0x00000001090438a0 @filter_argument=/^\/config\//>,
 #<SimpleCov::RegexFilter:0x0000000109043828 @filter_argument=/^\/db\//>,
 #<SimpleCov::StringFilter:0x0000000109042478 @filter_argument="/bin/">,
 #<SimpleCov::StringFilter:0x00000001090423d8 @filter_argument="/lib/tasks/auto_annotate_models.rake">,
 #<SimpleCov::StringFilter:0x0000000109042338 @filter_argument="/lib/tasks/coverage.rake">,
 #<SimpleCov::StringFilter:0x0000000109042298 @filter_argument="/spec/support/">,
 #<SimpleCov::StringFilter:0x00000001090421f8 @filter_argument="/spec/factories/">,
 #<SimpleCov::StringFilter:0x0000000109042158 @filter_argument="/spec/rails_helper.rb">,
 #<SimpleCov::StringFilter:0x00000001090420b8 @filter_argument="/spec/spec_helper.rb">,
 #<SimpleCov::BlockFilter:0x0000000109041fa0
  @filter_argument=#<Proc:0x0000000109042018 /Users/eoinkelly/Code/repos/rails-template/tmp/builds/tmpl_devise_app/spec/spec_helper.rb:29>>]
```

All seem to be in `@filters`.

The "root filter" [filters everything outside the project](https://github.com/simplecov-ruby/simplecov#default-root-filter-and-coverage-for-things-outside-of-it).

This filtering looks sensible to me so I think this is good to merge.

**Pull request merged**